spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/cassandra/CassandraDriverHealthIndicator.java

```diff
@@ -52,11 +52,18 @@ public CassandraDriverHealthIndicator(CqlSession session) {
 
 	@Override
 	protected void doHealthCheck(Health.Builder builder) throws Exception {
-		Row row = this.session.execute(SELECT).one();
-		builder.up();
-		if (row != null && !row.isNull(0)) {
-			builder.withDetail("version", row.getString(0));
+		Collection<Node> nodes = this.session.getMetadata().getNodes().values();
+		boolean atLeastOneUp = nodes.stream().map(Node::getState).anyMatch((state) -> state == NodeState.UP);
+		if (atLeastOneUp) {
+			builder.up();
 		}
+		else {
+			builder.down();
+		}
+
+		// fill details with version of the first node (if the version is not null)
+		nodes.stream().map(Node::getCassandraVersion).filter(Objects::nonNull).findFirst()
+				.ifPresent((version) -> builder.withDetail("version", version));
```

Comment:
What are the reasons for which a node wouldn't report a version? I think the first map operation should rather return the first node that's up and we should then use the information provided by that node.

Reply:
> What are the reasons for which a node wouldn't report a version?

A healthy Cassandra installation would always report its version. But we've seen some edge cases in the past: firstly, the `system.peers` table is subject to corruption (intentional or accidental); secondly, this may happen when a node is being upgraded. This is the reason why `com.datastax.oss.driver.api.core.metadata.Node#getCassandraVersion` is annotated with `@Nullable`.

Reply:
Thanks for the feedback @adutra. In the case we keep the node that states healthy, wouldn't that be correct to assume its version is set? Can you have a node reporting healthy and not providing a version at the same time?

Reply:
> Can you have a node reporting UP and not providing a version at the same time?

It is technically possible, although very unlikely.